### PR TITLE
Add architecture.md

### LINKED ISSUE #68: Improve architecture documentation
Move architecture docs from readme to a separate architecture.md file.

Show in detail the flow of requests for language features ie. diagnostics and quick fixes. From client to mp.ls to mp.jdt etc.

### PR DESCRIPTION
**Pull request by rzgry**
Fixes #68 

Moves architecture information to a separate architecture.md file.

Adds a walk through of microprofile-config.properties and Java file language features similar to https://github.com/redhat-developer/vscode-microprofile/blob/master/CONTRIBUTING.md#project-structure

Signed-off-by: Ryan Zegray <ryan.zegray@ibm.com>

### CODE REVIEW

File: docs/architecture.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# Architecture
+
+The language features provided by the [MicroProfile Language Server](./microprofile.ls) depend on the dependencies in the users Java Project.
```
Comment (kathrynkodama) [inline docs/architecture.md:3]: should be `user's Java project`

File: docs/architecture.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+For example, MicroProfile Fault Tolerance features (snippets, validation, completion, etc.) are only provided if the user's
+Java project has MicroProfile Fault Tolerance as a dependency.
+
+In order to avoid duplicating the work done by an IDEs Java Language Support (computing the classpath, rebuilding AST, etc.) the MicroProfile language server relies on an external component that understands and handles LSP command requests sent from the MicroProfile Language server.
```
Comment (kathrynkodama) [inline docs/architecture.md:7]: all instances of `MicroProfile Language Server` should be capitalized

File: docs/architecture.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+In order to avoid duplicating the work done by an IDEs Java Language Support (computing the classpath, rebuilding AST, etc.) the MicroProfile language Server relies on an external component that understands and handles LSP command requests sent from the MicroProfile Language Server.
+
+The MicroProfile Language Server has limited knowledge of the users Java project, and instead uses the external component for information on the Java project. In the context of Eclipse and VS Code, this external component is managed with [MicroProfile JDT](./microprofile.jdt).
+The external component does not necessarily require the JDT however, as different implementations can be used instead. For example [Quarkus Tools for IntelliJ](https://github.com/redhat-developer/intellij-quarkus) does not use the JDT and instead has its own implementation used to handle the requests from [`MicroProfileLanguageClientAPI`](https://github.com/eclipse/lsp4mp/blob/master/microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/api/MicroProfileLanguageClientAPI.java). In other words, the MicroProfile Language Server **does not** depend on JDT.
```
Comment (angelozerr) [inline docs/architecture.md:10]: @jeffmaury could you review this sentence please about IJ, if you think we should add another content. perhaps it should be fine to explain that you are using IJ Psi java API (I'm not sure if it's the right name).

File: docs/architecture.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+## microprofile-config.properties language features
+
+For properties completion, the MicroProfile Language Server uses the external component to detect all properties available on the classpath.
+
+For example:
+
+1.A MicroProfile project is opened and the IDE/editor sends a `textDocument/completion` request after completion has been triggered in the `microprofile-config.properties` file.
```
Comment (angelozerr) [inline docs/architecture.md:18]: 1. A

File: docs/architecture.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+The MicroProfile Language Server has limited knowledge of the users Java project, and instead uses the external component for information on the Java project. In the context of Eclipse and VS Code, this external component is managed with [MicroProfile JDT](./microprofile.jdt).
+The external component does not necessarily require the JDT however, as different implementations can be used instead. For example [Quarkus Tools for IntelliJ](https://github.com/redhat-developer/intellij-quarkus) does not use the JDT and instead has its own implementation used to handle the requests from [`MicroProfileLanguageClientAPI`](https://github.com/eclipse/lsp4mp/blob/master/microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/api/MicroProfileLanguageClientAPI.java). In other words, the MicroProfile Language Server **does not** depend on JDT.
+
+## microprofile-config.properties language features
+
+For properties completion, the MicroProfile Language Server uses the external component to detect all properties available on the classpath.
+
```
Comment (angelozerr) [inline docs/architecture.md:15]: I will add 

The external component compute and return those properties by following the same format than [Spring Boot Configuration Metadata](https://docs.spring.io/spring-boot/docs/current/reference/html/appendix-configuration-metadata.html#configuration-metadata-format).

```json
{
	"properties": [
		{
			"type": "java.lang.String",
			"extensionName": "microprofile-openapi-api",
			"required": true,
			"name": "mp.openapi.model.reader",
			"description": "Configuration property to specify the fully qualified name of the OASModelReader implementation."
		},
		{
			"type": "java.lang.String",
			"extensionName": "microprofile-openapi-api",
			"required": true,
			"name": "mp.openapi.filter",
			"description": "Configuration property to specify the fully qualified name of the OASFilter implementation."
		},
		{
			"type": "boolean",
			"defaultValue": "false",
			"extensionName": "microprofile-openapi-api",
			"required": false,
			"name": "mp.openapi.scan.disable",
			"description": "Configuration property to disable annotation scanning. Default value is `false`."
		},
		{
			"type": "java.lang.String",
			"extensionName": "microprofile-openapi-api",
			"required": true,
			"name": "mp.openapi.scan.packages",
			"description": "Configuration property to specify the list of packages to scan.\r\nFor example, `mp.openapi.scan.packages=com.xyz.PackageA,com.xyz.PackageB`."
		},
		{
			"type": "java.lang.String",
			"extensionName": "microprofile-openapi-api",
			"required": true,
			"name": "mp.openapi.scan.classes",
			"description": "Configuration property to specify the list of classes to scan.\r\nFor example, `mp.openapi.scan.classes=com.xyz.MyClassA,com.xyz.MyClassB`."
		},
		{
			"type": "java.lang.String",
			"extensionName": "microprofile-openapi-api",
			"required": true,
			"name": "mp.openapi.scan.exclude.packages",
			"description": "Configuration property to specify the list of packages to exclude from scans.\r\nFor example, `mp.openapi.scan.exclude.packages=com.xyz.PackageC,com.xyz.PackageD`."
		},
		{
			"type": "java.lang.String",
			"extensionName": "microprofile-openapi-api",
			"required": true,
			"name": "mp.openapi.scan.exclude.classes",
			"description": "Configuration property to specify the list of classes to exclude from scans.\r\nFor example, `mp.openapi.scan.exclude.classes=com.xyz.MyClassC,com.xyz.MyClassD`."
		},
		{
			"type": "java.lang.String",
			"extensionName": "microprofile-openapi-api",
			"required": true,
			"name": "mp.openapi.servers",
			"description": "Configuration property to specify the list of global servers that provide connectivity information.\r\nFor example, `mp.openapi.servers=https://xyz.com/v1,https://abc.com/v1`."
		},
		{
			"type": "java.lang.String",
			"extensionName": "microprofile-openapi-api",
			"required": true,
			"name": "mp.openapi.servers.path.{*}",
			"description": "Prefix of the configuration property to specify an alternative list of servers to service all operations in a path.For example, `mp.openapi.servers.path./airlines/bookings/{id}=https://xyz.io/v1`."
		},
		{
			"type": "java.lang.String",
			"extensionName": "microprofile-openapi-api",
			"required": true,
			"name": "mp.openapi.servers.operation.{*}",
			"description": "Prefix of the configuration property to specify an alternative list of servers to service an operation.Operations that want to specify an alternative list of servers must define an `operationId`, a unique string used to identify the operation.For example, `mp.openapi.servers.operation.getBooking=https://abc.io/v1`."
		},
		{
			"type": "java.lang.String",
			"extensionName": "microprofile-openapi-api",
			"required": true,
			"name": "mp.openapi.schema.{*}",
			"description": "Prefix of the configuration property to specify a schema for a specific class, in JSON format.The remainder of the property key must be the fully-qualified class name.The value must be a valid OpenAPI schema object, specified in the JSON format.The use of this property is functionally equivalent to the use of the `@Schema` annotationon a Java class, but may be used in cases where the application developer does not have access to the source code of a class."
		}
	]
}
```

Comment (angelozerr) [reply docs/architecture.md:15]: Perhaps in an another PR, but I think it should be nice to have a simple diagram with vscode completion which shows MP completion and LSP4MP server side which returns Spring Boot Metadata JSON list of properties.

File: docs/architecture.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+3. The IDE/editor receives the `microprofile/projectInfo` request and delegates it to the external component. (The [microprofile.jdt](./microprofile.jdt) for Eclipse/VSCode)
+
+4. The external component receives the command in the [MicroProfileDelegateCommandHandler#executeCommand](https://github.com/eclipse/lsp4mp/blob/master/microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/ls/MicroProfileDelegateCommandHandler.java#L54), computes [project information](https://github.com/eclipse/lsp4mp/blob/master/microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/commons/MicroProfileProjectInfo.java)(Project URI, the available MP config properties, etc.)
+
+5. The IDE / editor receives the project information and sends it back to the MicroProfile Language Server
+
+6. The MicroProfile Language Server receives the project information, adds it to its cache, and returns the completion options stored in the project information as the response to the original `textDocument/completion` request.
+
```
Comment (angelozerr) [inline docs/architecture.md:33]: Perhaps in an another PR but we should have a sequence diagram for that.

File: docs/architecture.md (hunk 3 above)
Comment (jeffmaury) [reply docs/architecture.md:10]: You can rewrite:
`does not use the JDT and instead has its own implementation used to handle the requests`
as
`does not use the JDT and instead has its own implementation based on [Program Structure Interface](https://jetbrains.org/intellij/sdk/docs/basics/architectural_overview/psi.html) used to handle the requests`

Comment (angelozerr) [reply docs/architecture.md:10]: Thanks @jeffmaury!

**Pull request merged**